src/main/java/ch/naviqore/service/Leg.java

```diff
@@ -0,0 +1,22 @@
+package ch.naviqore.service;
+
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
+
+public interface Leg {
+    @NotNull Location getDepartureLocation();
+
+    @NotNull Location getArrivalLocation();
+
+    @Nullable Stop getDepartureStop();
+
+    @Nullable Stop getArrivalStop();
+
+    @NotNull DepartureTime getDepartureTime();
+
+    @NotNull ArrivalTime getArrivalTime();
+
+    @Nullable Trip getTrip();
```

Comment:
Maybe we could split Legs in `WalkingLeg` and `RouteLeg`? Then have an order Integer which stores the position in the complete connection? Somehow I feel like avoiding nullable fields, which introduce a lot of null checks and potentially NullPointerExceptions?

I not sure yet what the alternative could look like though...

Reply:
I now it's suboptimal but I couldn't think of any other way to handle it. All alternatives will also always require some kind of extra logic.